### PR TITLE
feat(prover): introduce `TierGuardianMinority`

### PR DESCRIPTION
**Pull request by davidtaikocha**

### CODE REVIEW

File: internal/testutils/suite.go [hunk 1]
```diff
@@ -56,14 +56,15 @@ func (s *ClientTestSuite) SetupTest() {
...
-		GuardianProverAddress: common.HexToAddress(os.Getenv("GUARDIAN_PROVER_CONTRACT_ADDRESS")),
-		L2EngineEndpoint:      os.Getenv("L2_EXECUTION_ENGINE_AUTH_ENDPOINT"),
-		JwtSecret:             string(jwtSecret),
+		L1Endpoint:                    os.Getenv("L1_NODE_WS_ENDPOINT"),
+		L2Endpoint:                    os.Getenv("L2_EXECUTION_ENGINE_WS_ENDPOINT"),
+		TaikoL1Address:                common.HexToAddress(os.Getenv("TAIKO_L1_ADDRESS")),
+		TaikoL2Address:                common.HexToAddress(os.Getenv("TAIKO_L2_ADDRESS")),
+		TaikoTokenAddress:             common.HexToAddress(os.Getenv("TAIKO_TOKEN_ADDRESS")),
+		GuardianProverMajorityAddress: common.HexToAddress(os.Getenv("GUARDIAN_PROVER_CONTRACT_ADDRESS")),
+		GuardianProverMinorityAddress: common.HexToAddress(os.Getenv("GUARDIAN_PROVER_CONTRACT_ADDRESS")),
```
[the remark this answers is not in the document]
Comment (YoGhurt111) [reply internal/testutils/suite.go:65]: ok

File: prover/config.go [hunk 2]
```diff
@@ -101,7 +102,7 @@ func NewConfigFromCliContext(c *cli.Context) (*Config, error) {
 	}
 
 	// If we are running a guardian prover, we need to prove unassigned blocks and run in contester mode by default.
-	if c.IsSet(flags.GuardianProver.Name) {
+	if c.IsSet(flags.GuardianProverMajority.Name) && c.IsSet(flags.GuardianProverMinority.Name) {
```
[the remark this answers is not in the document]
Comment (YoGhurt111) [reply prover/config.go:105]: If we are running a guardian prover, we need to set the `GuardianProverMajority` and `GuardianProverMinority` flags both. If one of them is missed, we assume that it is not running as a guardian prover, as in the following logic.

File: prover/proof_producer/minority_guardian_producer.go [hunk 3]
```diff
@@ -0,0 +1,59 @@
+package producer
```
[the remark this answers is not in the document]
Comment (YoGhurt111) [reply prover/proof_producer/minority_guardian_producer.go:1]: ok

File: prover/proof_submitter/transaction/builder.go [hunk 4]
```diff
@@ -73,8 +78,13 @@ func (a *ProveBlockTxBuilder) Build(
...
 			}
 		} else {
-			to = a.guardianProverAddress
-
+			if tier > encoding.TierGuardianMinorityID {
+				to = a.majorityGuardianProverAddress
+			} else if tier == encoding.TierGuardianMinorityID && a.minorityGuardianProverAddress != ZeroAddress {
+				to = a.minorityGuardianProverAddress
+			} else {
+				return nil, fmt.Errorf("tier %d need set minorityGuardianProverAddress", tier)
```
Comment (mask-pp) [inline prover/proof_submitter/transaction/builder.go:86]: If `tier < encoding.TierGuardianMinorityID` and `a.majorityGuardianProverAddress ~= common.Address{}` then will return error message, but this error content is not correct.

**Review by mask-pp — APPROVED** [review]
LGTM, just a carp

File: prover/proof_submitter/transaction/builder.go (hunk 4 above)
Comment (YoGhurt111) [reply prover/proof_submitter/transaction/builder.go:86]: The switch case only happens in `guardian` status and `guardian = tier >= encoding.TierGuardianMinorityID`.

File: pkg/rpc/client.go [hunk 5]
```diff
@@ -27,27 +27,29 @@ type Client struct {
...
 	L1Beacon *BeaconClient
 	// Protocol contracts clients
-	TaikoL1        *bindings.TaikoL1Client
-	TaikoL2        *bindings.TaikoL2Client
-	TaikoToken     *bindings.TaikoToken
-	GuardianProver *bindings.GuardianProver
+	TaikoL1                *bindings.TaikoL1Client
+	TaikoL2                *bindings.TaikoL2Client
+	TaikoToken             *bindings.TaikoToken
+	MajorityGuardianProver *bindings.GuardianProver
```
Comment (davidtaikocha) [inline pkg/rpc/client.go:33]:
```suggestion
	GuardianProverMajority *bindings.GuardianProver
```

File: pkg/rpc/client.go [hunk 6]
```diff
@@ -27,27 +27,29 @@ type Client struct {
...
+	MinorityGuardianProver *bindings.GuardianProver
```
Comment (davidtaikocha) [inline pkg/rpc/client.go:34]:
```suggestion
	GuardianProverMinority *bindings.GuardianProver
```

File: pkg/rpc/client.go [hunk 7]
```diff
@@ -110,16 +112,22 @@ func NewClient(ctx context.Context, cfg *ClientConfig) (*Client, error) {
 	}
 
 	var (
-		taikoToken     *bindings.TaikoToken
-		guardianProver *bindings.GuardianProver
+		taikoToken             *bindings.TaikoToken
+		majorityGuardianProver *bindings.GuardianProver
```
Comment (davidtaikocha) [inline pkg/rpc/client.go:116]:
```suggestion
		guardianProverMajority *bindings.GuardianProver
```

File: pkg/rpc/client.go [hunk 8]
```diff
@@ -110,16 +112,22 @@ func NewClient(ctx context.Context, cfg *ClientConfig) (*Client, error) {
...
+		minorityGuardianProver *bindings.GuardianProver
```
Comment (davidtaikocha) [inline pkg/rpc/client.go:117]:
```suggestion
		guardianProverMinority *bindings.GuardianProver
```

File: prover/config.go [hunk 9]
```diff
@@ -116,8 +117,12 @@ func NewConfigFromCliContext(c *cli.Context) (*Config, error) {
 		if !c.IsSet(flags.L2NodeVersion.Name) {
 			return nil, errors.New("--prover.l2NodeVersion flag is required if guardian prover is set")
 		}
-	}
 
+		// If we are running a guardian prover, a raiko host endpoint is required.
+		if !c.IsSet(flags.RaikoHostEndpoint.Name) {
```
Comment (davidtaikocha) [inline prover/config.go:122]: We can remove this check now, since https://github.com/taikoxyz/taiko-client/pull/798

File: prover/init.go [hunk 10]
```diff
@@ -113,7 +113,16 @@ func (p *Prover) initProofSubmitters(
 				ProofType:         proofProducer.ProofTypeSgx,
 				Dummy:             p.cfg.Dummy,
 			}
-		case encoding.TierGuardianID:
+		case encoding.TierGuardianMinorityID:
```
Comment (davidtaikocha) [inline prover/init.go:116]: Can we just reuse `proofProducer.NewGuardianProofProducer` here?

**Pull request merged**